src/bloom/hierarchical.rs

```diff
@@ -0,0 +1,335 @@
+use std::iter::FromIterator;
+
+use super::*;
+
+/// Container for types which is support bloom filtering
+#[derive(Debug)]
+pub struct HierarchicalBloom<Child>
+where
+    Child: BloomProvider + Send + Sync,
+{
+    inner: HierarchicalBloomInner,
+    children: Vec<Child>,
+    group_size: usize,
+}
+
+#[derive(Debug)]
+enum HierarchicalBloomInner {
+    /// Inner filters
+    Vec(Option<Bloom>, Vec<HierarchicalBloomInner>),
+    /// Leaf
+    Item(usize),
+}
+
+impl Default for HierarchicalBloomInner {
+    fn default() -> Self {
+        Self::Vec(Default::default(), Default::default())
+    }
+}
+
+impl<Child> Default for HierarchicalBloom<Child>
+where
+    Child: BloomProvider + Send + Sync,
+{
+    fn default() -> Self {
+        Self {
+            inner: Default::default(),
+            children: Default::default(),
+            group_size: 8,
+        }
+    }
+}
+
+impl HierarchicalBloomInner {
```

Comment:
There should also be a method, that allows to get the list of BLOBs that may contain the key. It is not enough just to say, that some BLOB has specified key. With hierarchical filters we can reduce the number of BLOBs to scan by rejecting the BLOB groups that definitely do not contain the key

Reply:
Done in new version